platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,495 @@
+is not represented in protobuf or transmitted through gossip. One field populated in the metadata will be the byte[]
+that represents the protobuf-serialized form of the `EventCore`. Since gossip read this from the wire, there is no work
+to produce this `byte[]`, and since the `byte[]` will be needed for Gossip, we need to keep it anyway. But we will also
+hash this `byte[]` to produce the event hash, and store this hash in the metadata.
+
+After hashing, Event Intake will deduplicate events. It has a hash->event map allowing it to cheaply verify whether the
+event is a duplicate. If it is, the duplicate is discarded. Otherwise, the event is checked for "syntactic" correctness.
+For example, are all required fields populated, etc. While the Gossip system has already checked to ensure the payload
+of the event (its transactions) are limited in size and count, Event Intake will also check this as a safety measure in
+the event of bugs (if this check fails here, a noisy log statement should be produced, since this should never happen).
+
+If an event is valid, then we finally check the signature. Since validation and deduplication and hashing are
+significantly less expensive than signature verification, we wait on signature verification until the other steps are
+completed. The operating principle is that we want to fail fast and limit work for further stages in the pipeline.
+
+If an event has a very old birth-round that is ancient, it is dropped. If a node sends a large number of ancient events,
+it may end up being disciplined (the exact rules around this will be defined in subsequent design docs for the
+Event Intake module).
+
+#### Topological Ordering
+
+Events are buffered if necessary to ensure that each parent event has been emitted from Event Intake before any child
+events. A simple map (the same used for deduplication) can be used here. Given some event, for each parent, look up the
+parent by its hash. If each parent is found in the map, then emit the event. Otherwise, remember the event so when the
+missing parent is received, the child may be emitted. The current implementation uses what is known as the "orphan
+buffer" for this purpose.
+
+Since Event Intake will also maintain some buffers, it needs to know about the progression of the hashgraph,
+so it can evict old events. In this case, the "orphan buffer" holds events until either the parent events have
+arrived, or the events have become ancient due to the advancement of the "non-ancient event window" and the event is
+dropped from the buffer. This document does not prescribe the existence of the orphan buffer or the method by which
+events are sorted and emitted in topological order, but it does describe a method by which old events can be dropped.
+
+#### Birth-Round Filtering
+
+When an event is created, it is assigned a "birth round". This is the most recent consensus round on the node at the
```

Comment:
I was wondering whether we use the most-recent consensus round, or the just next round (the one that voted on the consensus round), or the one after that (the one that collected the votes). I think your argument actually would suggest that we should let the birth round be the most recent "vote collecting round".

Reply:
One thing to consider, is that the roster used for verification is not necessarily the same one used for consensus. For example, suppose we just came to consensus on round 100 and we put the new roster to be active in round 104. And then suppose we create a new event with birth round 100. When it gets gossiped, every other node will use the address book (R0) associated with round 100 for that event, but of course, the event will probably end up in round 101 or newer (I would guess most likely much newer).

Or we could give the birth round as 101 and it will still use roster R0 because roster R0 was used for round 101 and 102 (remember that 3 rounds must exist to come to consensus on the first). And that event will still most likely come to consensus in some round after 102, maybe even after 104.

Basically, we have no guarantee of any correlation between the birth round, and the round that will be used for voting, or the final consensus round the event is included in. For that reason, keeping the birth-round as the most-recent consensus round (round that came to consensus) seems reasonable.

Reply:
What is a `vote collecting round`?   In virtual voting there are "future rounds" created to determine the judges and famous witnesses, but those future round assignments on events are erased once the `pending consensus round` comes to consensus.   I also think its possible that judges come from many "future rounds" into the future, and not just the next round in the future. 

I think we should consider a generate case.  Roster B is the next roster after Roster A and there is > 1/3 consensus weight of a delta between the two rosters.  As consensus is working towards the round on which Roster B is adopted, when does the consensus algorithm fail to advance?   I'll ask this question in the consensus channel.

Reply:
@edward-swirldslabs Is this question cleared up from your perspective?

Reply:
yes